File: mobile/packages/footprint-react-native/src/onboarding-components/queries/save.ts
import { AUTH_HEADER } from '@onefootprint/types';
import request from 'src/utils/request';

import type { SaveDataRequest, SaveDataResponse } from '../types/save-request-data';

const getDataKind = (data: Record<string, unknown>) => {
  const hasId = Object.entries(data).some(([key]) => key.startsWith('id.'));
  const hasBusiness = Object.entries(data).some(([key]) => key.startsWith('business.'));
  return { hasId, hasBusiness };
};

const save = async (payload: SaveDataRequest) => {
  const { data } = payload;
  if (!Object.entries(data).length) {
    // If there's no data to send to the backend, short circuit
    return {} as SaveDataResponse;
  }

  const { hasId, hasBusiness } = getDataKind(data);
  if (hasId && hasBusiness) {
    throw new Error("You can't submit id and business at the same time");
  }

  const url = hasBusiness ? '/hosted/business/vault' : '/hosted/user/vault';
  const response = await request<SaveDataResponse>({
    url,
    method: 'PATCH',
    data,
    headers: {
      [AUTH_HEADER]: payload.authToken,
    },
    disableCaseConverter: true,
  });
  return response;
};

export default save;

File: mobile/packages/footprint_flutter/lib/src/onboarding-components/queries/save.dart
import 'dart:convert';

import 'package:footprint_flutter/src/config/constants.dart';
import "package:footprint_flutter/src/onboarding-components/models/save_data_request.dart";
import 'package:http/http.dart' as http;

typedef DataKind = ({bool hasId, bool hasBusiness});

DataKind getDataKind(Map<String, dynamic> data) {
  final hasId = data.entries.any((entry) => entry.key.startsWith('id.'));
  final hasBusiness =
      data.entries.any((entry) => entry.key.startsWith('business.'));
  return (hasBusiness: hasBusiness, hasId: hasId);
}

Future<SaveDataResponse> save(SaveDataRequest payload) async {
  final data = payload.data;
  if (data.entries.isEmpty) {
    return SaveDataResponse(data: null);
  }

  final dataKind = getDataKind(data);
  if (dataKind.hasId && dataKind.hasBusiness) {
    throw Exception("You can't submit id and business at the same time");
  }

  final url =
      dataKind.hasBusiness ? '/hosted/business/vault' : '/hosted/user/vault';
  var response = await http.patch(
    Uri.parse('$apiBaseUrl$url'),
    headers: {
      AUTH_HEADER: payload.authToken,
      'Content-Type': 'application/json'
    },
    body: jsonEncode(data),
  );

  if (response.statusCode == 200) {
    return SaveDataResponse(data: response.body);
  } else {
    throw Exception('Failed to save data. Error: ${response.body}');
  }
}
